more performance tweaks

seebs opened the pull request:
This got me from about 24.8fps to about 34.9fps for my test data. All reasonably clean, I hope.

faiface commented on geometry.go:
```diff
@@ -49,6 +49,21 @@ func (u Vec) XY() (x, y float64) {
+	}
+	return Vec{X: length * u.X, Y: length * u.Y}
+}
+
```
I dislike this function. Duplicates code, this is something we should leave to the inlining capabilities of the compiler (which improve over time).

faiface commented on imdraw/imdraw.go:
```diff
@@ -128,7 +128,9 @@ func (imd *IMDraw) Draw(t pixel.Target) {
-	imd.Color = pixel.ToRGBA(imd.Color)
+	if _, ok := imd.Color.(pixel.RGBA); !ok {
+		imd.Color = pixel.ToRGBA(imd.Color)
+	}
```
Again, this duplicates code, should be left to the compiler.

seebs replied:
it can't be. The problem here is that pixel.ToRGBA returns a pixel.RGBA, which *implements*, but *isn't*, a color.Color interface. So the compiler has to box the result, which is an object allocation, which is why this is so unexpectedly effective.

faiface commented on pixelgl/gltriangles.go:
```diff
@@ -101,15 +101,16 @@ func (gt *GLTriangles) updateData(t pixel.Triangles) {
 
 	// TrianglesData short path
+	s := gt.vs.Stride()
+	l := gt.Len()
```
Use longer variable names here. These variables are used in a larger context than 5 lines of code, so this is hurts readability.

seebs replied:
good point.

seebs commented:
Updated. Fixed the variable names, split out the Normal code. If we ever want to get a bunch of the remaining cost out of the Normal stuff, probably math.Sqrt() is good enough; right now, math.Hypot() is chewing up more time than all the rest of the normal calcs put together.

faiface commented on geometry.go:
```diff
@@ -49,6 +49,15 @@ func (u Vec) XY() (x, y float64) {
+	}
+	return Vec{X: u.Y / l, Y: -u.X / l}
+}
+
```
Ok, all is good except one thing. This implementation is inconsistent with Unit(). On a zero sized vector, Unit() returns (1,0), but this returns (0,0). No good, this should return (0,1) in such case.

Also, is using u.Len() here too expensive?

seebs commented:
Okay, updated. I added the To() thing because polyline's easier (to me, anyway?) to read when p0->p1 is written as p0.To(p1) instead of p1.Sub(p0). Also fixed the comment.

faiface commented on geometry.go:
```diff
@@ -49,6 +49,11 @@ func (u Vec) XY() (x, y float64) {
 	return u.X, u.Y
 }
 
+// Normal returns a vector normal to u (rotated by math.pi/2)
```
missing period and I'd say:
```
// Normal returns a vector normal to u. Equivalent to u.Rotate(math.Pi / 2), but faster.
```

faiface commented on geometry.go:
```diff
@@ -65,6 +70,14 @@ func (u Vec) Sub(v Vec) Vec {
 	}
 }
 
+// To returns the vector from vector u to vector v, equivalent to v.Sub(u).
```
```go
// To returns the vector from u to v. Eequivalent to v.Sub(u).
```

vector was too repetitive
